pkg/merkletree/tree_test.go

```diff
@@ -0,0 +1,357 @@
+	})
+
+	t.Run("wrong proof", func(t *testing.T) {
+		t.Parallel()
+		root := []byte("036")
+		proof := [][]byte{[]byte("001"), []byte("007"), []byte("026")}
+		leaf := []byte("001")
+		require.False(t, VerifyProof(root, proof, leaf, mockHashFunc))
+	})
+}
+
+func TestProperties(t *testing.T) {
+	t.Parallel()
+
+	// We test the properties in a bunch of hash functions to make sure the
+	// kind of hash function has no influence on properties.
+	hashFuncs := []func(...[]byte) []byte{
+		nil,
+		mockHashFunc,
+		crypto.Keccak256,
+		crypto.Keccak512,
+		func(b ...[]byte) []byte {
+			h := sha3.Sum224(b[0])
+			return h[:]
+		},
+		func(b ...[]byte) []byte {
+			h := sha3.Sum256(b[0])
+			return h[:]
+		},
+		func(b ...[]byte) []byte {
+			h := sha3.Sum384(b[0])
+			return h[:]
+		},
+		func(b ...[]byte) []byte {
+			h := sha3.Sum512(b[0])
+			return h[:]
+		},
+		func(b ...[]byte) []byte {
+			h := sha1.Sum(b[0])
+			return h[:]
+		},
+	}
+
+	t.Run("tree holds merkle tree property", func(t *testing.T) {
+		t.Parallel()
+
+		for _, hashFunc := range hashFuncs {
+			property := func(leaves [][]byte) bool {
+				if len(leaves) == 0 {
+					return true
+				}
+
+				tree, err := NewTree(leaves, hashFunc)
+				if err != nil {
+					return false
+				}
+
+				return tree.VerifyTree()
+			}
+			require.NoError(t, quick.Check(property, nil))
+		}
+	})
+
+	t.Run("leaves are always sorted", func(t *testing.T) {
+		t.Parallel()
+
+		for _, hashFunc := range hashFuncs {
+			property := func(leaves [][]byte) bool {
+				if len(leaves) == 0 {
+					return true
+				}
+
+				tree, err := NewTree(leaves, hashFunc)
+				if err != nil {
+					return false
+				}
+
+				return sort.SliceIsSorted(tree.leaves, func(i, j int) bool {
+					return bytes.Compare(tree.leaves[i].hash, tree.leaves[j].hash) == -1
+				})
+			}
+			require.NoError(t, quick.Check(property, nil))
+		}
+	})
+
+	t.Run("height of the tree is correct", func(t *testing.T) {
+		t.Parallel()
+		for _, hashFunc := range hashFuncs {
+			property := func(leaves [][]byte) bool {
+				if len(leaves) == 0 {
+					return true
+				}
+
+				tree, err := NewTree(leaves, hashFunc)
+				if err != nil {
+					return false
+				}
+
+				return heightOfTree(tree.root) == expectedHeight(len(leaves))
+			}
+			require.NoError(t, quick.Check(property, nil))
+		}
+	})
+
+	t.Run("if number of leaves is odd, then the last leaf is duplicated", func(t *testing.T) {
+		t.Parallel()
+		for _, hashFunc := range hashFuncs {
+			property := func(leaves [][]byte) bool {
+				if len(leaves)%2 == 0 {
+					return true
+				}
+
+				tree, err := NewTree(leaves, hashFunc)
+				if err != nil {
+					return false
+				}
+
+				return len(tree.leaves) == len(leaves)+1 &&
+					bytes.Equal(tree.leaves[len(tree.leaves)-1].hash, tree.leaves[len(tree.leaves)-2].hash)
+			}
+			require.NoError(t, quick.Check(property, nil))
+		}
+	})
+
+	t.Run("every leaf proof is correctly verifiable", func(t *testing.T) {
```

Comment:
This tests that every leaf can be correctly verified by getting the proof of each one and verifying